File: check_balanced.py

class Node(object):
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None

	def __repr__(self):
		return "({})".format(self.data)

def check_balanced(node):
	if not node:
		return 0
	left = check_balanced(node.left)
	if left == -1:
		return -1
	right = check_balanced(node.right)
	if right == -1:
		return -1
	if abs(left - right) > 1:
		return -1
	return max(left, right) + 1

def main():
	root = Node(1)
	a = Node(2)
	b = Node(3)
	c = Node(4)
	d = Node(5)
	e = Node(6)
	f = Node(7)
	g = Node(8)
	h = Node(9)
	i = Node(10)
	j = Node(11)
	k = Node(12)
	root.left = a
	root.right = b
	b.left = c
	c.left = d
	c.right = j
	b.right = e
	a.left = h
	a.right = k
	h.left = i
	print(check_balanced(root))

if __name__ == "__main__":
	main()

File: check_balanced.cpp
#include <iostream>

using namespace std;

template <class T>
class Node{
public:
	Node();
	~Node();
	T data;
	Node<T>* left;
	Node<T>* right;
};

template <class T>
Node<T>::Node(){
	left = NULL;
	right = NULL;
}

template <class T>
Node<T>::~Node(){
	delete data;
}

int checkBalanced(Node<int>* node){
	if(!node){
		return 0;
	}
	int left = checkBalanced(node->left);
	if(left == -1){
		return -1;
	}
	int right = checkBalanced(node->right);
	if(right == -1){
		return -1;
	}
	if(abs(left - right) > 1){
		return -1;
	}
	return max(left, right) + 1;
}

int main(){

	Node<int>* root = new Node<int>();
	root->data = 2;
	Node<int>* a = new Node<int>();
	a->data = 3;
	Node<int>* b = new Node<int>();
	b->data = 4;
	Node<int>* c = new Node<int>();
	c->data = 5;
	Node<int>* d = new Node<int>();
	d->data = 6;
	Node<int>* e = new Node<int>();
	e->data = 7;
	Node<int>* f = new Node<int>();
	f->data = 8;
	Node<int>* g = new Node<int>();
	g->data = 9;
	root->left = a;
	root->right = b;
	a->left = c;	
	a->right = e;	
	a->left->left = g;
	b->left = f;
	b->right = d;

	cout << checkBalanced(root) << endl;

	return 0;
}
